File: skills/code-architect/references/code-examples/cpp/Adapter.cpp
// Adapter — make an incompatible interface usable via a wrapper.
#include <iostream>
#include <memory>
#include <string>

// Target interface the client expects.
class Endpoint {
public:
    virtual ~Endpoint() = default;
    virtual std::string send(const std::string& payload) const = 0;
};

// Incompatible existing service with a different signature.
class LegacyService {
public:
    std::string transmit(const std::string& blob, int channel) const {
        return "legacy[ch" + std::to_string(channel) + "]:" + blob;
    }
};

// Adapter makes LegacyService usable through the Endpoint interface.
class LegacyAdapter : public Endpoint {
    LegacyService service;
    int channel;
public:
    explicit LegacyAdapter(int ch) : channel(ch) {}
    std::string send(const std::string& payload) const override {
        return service.transmit(payload, channel);
    }
};

void client(const Endpoint& e) {
    std::cout << e.send("hello") << "\n";
}

int main() {
    LegacyAdapter adapter{7};
    client(adapter);
    return 0;
}

File: skills/code-architect/references/code-examples/ruby/adapter.rb
# frozen_string_literal: true

# Adapter — make an incompatible interface usable via a wrapper.

# Target interface the client expects: #area.
class Client
  def describe(shape) = "area = #{format('%.2f', shape.area)}"
end

# Adaptee with an incompatible API (returns dimensions, no #area).
class LegacyRectangle
  attr_reader :width, :height

  def initialize(width, height)
    @width = width
    @height = height
  end
end

# Adapter wraps the adaptee and exposes the target interface.
class RectangleAdapter
  def initialize(legacy) = @legacy = legacy

  def area = @legacy.width * @legacy.height
end

# A native shape already conforming to the target.
class Disc
  def initialize(radius) = @radius = radius
  def area = Math::PI * @radius**2
end

if __FILE__ == $PROGRAM_NAME
  client = Client.new
  legacy = LegacyRectangle.new(3, 4)

  puts client.describe(RectangleAdapter.new(legacy))
  puts client.describe(Disc.new(2))
end
